syntax = "proto3";

package user_service;

service UserService {
    rpc Register (RegisterRequest) returns (RegisterResponse);
    rpc AuthenticatePhoneNumber (AuthenticatePhoneNumberRequest) returns (AuthenticatePhoneNumberResponse);
    rpc Login (LoginRequest) returns (LoginResponse);
    rpc GetUserInfo (GetUserInfoRequest) returns (GetUserInfoResponse);
    rpc AddAddress (AddAddressRequest) returns (AddressResponse);
    rpc ModifyAddress (ModifyAddressRequest) returns (AddressResponse);
    rpc DeleteAddress (DeleteAddressRequest) returns (AddressResponse);
    rpc SetPreferredAddress (SetPreferredAddressRequest) returns (AddressResponse);
}

message RegisterRequest {
    string first_name = 1;
    string last_name = 2;
    string phone_number = 3;
    string password = 4;
}

message RegisterResponse {
    string auth_code = 1;
}

message AuthenticatePhoneNumberRequest {
    string phone_number = 1;
    string auth_code = 2;
}

message AuthenticatePhoneNumberResponse {
    bool success = 1;
}

message LoginRequest {
    string phone_number = 1;
    string password = 2;
}

message LoginResponse {
    string token = 1;
}

message GetUserInfoRequest {
    string user_id = 1;
}

message GetUserInfoResponse {
    string first_name = 1;
    string last_name = 2;
    string phone_number = 3;
    string email = 4;
    bool phone_number_verified = 5;
    string gender = 6;
    string created_at = 7;
    string updated_at = 8;
}

message AddAddressRequest {
    string user_id = 1;
    string address_line_1 = 2;
    string address_line_2 = 3;
    string city = 4;
    string state = 5;
    string postal_code = 6;
    string country = 7;
}

message AddressResponse {
    string id = 1;
    bool success = 2;
}

message ModifyAddressRequest {
    string address_id = 1;
    string address_line_1 = 2;
    string address_line_2 = 3;
    string city = 4;
    string state = 5;
    string postal_code = 6;
    string country = 7;
}

message DeleteAddressRequest {
    string address_id = 1;
}

message SetPreferredAddressRequest {
    string address_id = 1;
}
